include/circt/Dialect/FIRRTL/FIRRTLOps.h

```diff
@@ -174,6 +175,99 @@ struct FirMemory {
   }
   StringAttr getFirMemoryName() const;
 };
+
+// Record of the inner sym name, the module name and the corresponding
+// operation. Also the port index, if the symbol is on a module port.
+struct InnerRefRecord {
+  mlir::StringAttr mod, innerSym;
+  mlir::Operation *op = nullptr;
+  unsigned portIdx = 0;
+  InnerRefRecord(StringAttr mod, StringAttr innerSym, Operation *op)
+      : mod(mod), innerSym(innerSym), op(op) {}
+  InnerRefRecord(StringAttr mod, StringAttr innerSym, Operation *op,
+                 unsigned portIdx)
+      : mod(mod), innerSym(innerSym), op(op), portIdx(portIdx) {}
+  InnerRefRecord(hw::InnerRefAttr ref)
+      : mod(ref.getModule()), innerSym(ref.getName()) {}
+  bool operator<(const InnerRefRecord &rhs) const {
+    return (innerSym.getValue() < rhs.innerSym.getValue() ||
+            (innerSym == rhs.innerSym && mod.getValue() < rhs.mod));
+  }
+  bool operator==(const InnerRefRecord &rhs) const {
+    return (innerSym == rhs.innerSym && mod == rhs.mod);
+  }
+};
+// A list of inner sym and the corresponding operations. Can be used for
+// efficient insertion and then searching. It can be used when the pattern is
+// first all the records are inserted, and then searched for. After insertion,
+// the list must be sorted, and then binary search is used.
+struct InnerRefList {
```

Comment:
This should conform to a map or set compatible interfaces.  I don't think it should block this PR, but when the structure is generalized we can clean it up.